### PR TITLE
feat: add var promtail_binary_local_dir 

### PR DESCRIPTION
**Pull request by OuiSouss**
# Feat : 
- Add var promtail_binary_local_dir in order to use directly binary of promtail when download from internet is not possible
- Add var promtail_skip_install in order to skip installation task when promtail is already installed

### CODE REVIEW

**Comment by patrickjahns** [comment]
@OuiSouss 
Thank you for the contribution.

As this changes the behaviour of the role - it would be great to have a scenario covering the different path of using the feature. 
Have you thought of yet, what would be the best approach to add tests for your use-case?

**Comment by OuiSouss** [comment]
Hello, 

About this question : 
> Have you thought of yet, what would be the best approach to add tests for your use-case?

I haven't thought of a test case yet because I haven't mastered testing with ansible.
But, I presume that it is necessarary to test when the variable  promtail_binary_local_dir is not empty, that the directory {{ promtail_install_dir }}/{{ promtail_version }} is well created and the next task has copied the file at the good destination.

About this question : 
> can you elaborate on the use-case where it makes sense to skip installation?

For example, my use case for this variable is in the case, where I have already installed  and I know it is installed but I just want to change the configuration. And in the case where I made a mistake on this configuration, I just want to change it back to a normal configuration without necessarily reinstalling the tool because it is well installed.

I don't know if I'm expressing myself correctly. If not, please tell me so that I will try to rephrase it

**Comment by patrickjahns** [comment]
> For example, my use case for this variable is in the case, where I have already installed and I know it is installed but I just want to change the configuration. And in the case where I made a mistake on this configuration, I just want to change it back to a normal configuration without necessarily reinstalling the tool because it is well installed.

Can you clarify, if in that use-case promtail was installed via the role or not? 
As mentioned previously, the role is based on the premise, that it will install promtail and manage it thereafter ( updates and configuration changes ). It is not intended to be used with externally installed promtail versions. For that specific scenario, a lightweight role only managing the configuration ( and eventually restarting a service ) should be enough.

The role is designed idempotent, so that any successive runs should not execute the download/installation steps ( but ansible will show "no changes" ) so that you can update the configuration at any point in time and the suggested variable is not required

> I haven't thought of a test case yet because I haven't mastered testing with ansible.
But, I presume that it is necessarary to test when the variable promtail_binary_local_dir is not empty, that the directory {{ promtail_install_dir }}/{{ promtail_version }} is well created and the next task has copied the file at the good destination.

Can I ask that you write a scenario for testing the functionality of `promtail_binary_local_dir` - otherwise it will be hard to maintain this switch in the long run and ensure that it will not break when changes are required. In https://github.com/patrickjahns/ansible-role-promtail/tree/master/molecule you can see 2 simple test matrix scenarios and adapt them to testing your specific addition

**Comment by patrickjahns** [comment]
I've given the concept of `promtail_binary_local_dir` another thought. Wouldn't it make sense to just provide a path to the zip file, so that the role performs as before, just without the download?

**Comment by OuiSouss** [comment]
> The role is designed idempotent, so that any successive runs should not execute the download/installation steps ( but ansible will show "no changes" ) so that you can update the configuration at any point in time and the suggested variable is not required

Yes, i forgot the aspect of indempotence. Indeed, this variable is useless. I'll remove it.

> I've given the concept of promtail_binary_local_dir another thought. Wouldn't it make sense to just provide a path to the zip file, so that the role performs as before, just without the download?
Can you clarify what the added benefit of using the extracted binary is?

Exactly, it might be more logical to keep the archive, it would remove a manual task.
So, in this case we just need to add a `when` condition on `Download promtail binaries` task to check if the archive `{{ promtail_tmp_dir }}/{{ promtail_version }}_promtail-linux-{{ go_arch }}.zip`  exist and skip when the archive is already downloaded. Is this a best alternative before I modify the code ?

**Comment by patrickjahns** [comment]
If you can place the archive in the configured `"{{ promtail_tmp_dir }}/{{ promtail_version }}_promtail-linux-{{ go_arch }}.zip"` path, it will already skip downloading if the file exists. See: https://github.com/patrickjahns/ansible-role-promtail/blob/master/tasks/install.yml#L61-L67

Do I understand you correctly - the main idea is, that you want to avoid that the role is required to fetch information from the internet - correct?

So we need to:
- Prevent the preflight requests from fetching information from the web https://github.com/patrickjahns/ansible-role-promtail/blob/master/tasks/preflight.yml 
- If the path on the system for the zip file needs to be configurable, add a variable to do so and skip the download accordingly

**Comment by patrickjahns** [comment]
@OuiSouss 
Are you still interested in bringing this feature into the role?

**Comment by patrickjahns** [comment]
@OuiSouss 
Thank you for the feedback - no worries - take your time. 

> I worked this weekend on testing with molecule. But, I still have problems with last task, when testing if the service is up.
> I will commit and push this work. Then we can discuss about it.

Feel free to also push unfinished work to share and iterate on it together

**Comment by OuiSouss** [comment]
I found a way to test the systemd services but changing the image from ubuntu to centos. I don't know if this is a problem.

File: molecule/scenario_with_local_binary_directory/molecule.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+lint: |
+  set -e
+  yamllint .
+  flake8
+
+platforms:
+  - name: instance
+    image: quay.io/centos/centos:stream8
+    command: /sbin/init
+    privileged: True
```
Comment (patrickjahns) [inline molecule/scenario_with_local_binary_directory/molecule.yml:15]: Can you elaborate why the change to a different image is required, compared to the default setup of the role in i.e.
https://github.com/patrickjahns/ansible-role-promtail/blob/master/molecule/default/molecule.yml#L12-L19 ?

File: molecule/scenario_with_local_binary_directory/converge.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+- name: Converge
+  hosts: all
+  tasks:
+    - name: "Include ansible-role-promtail"
+      include_role:
+        name: "ansible-role-promtail"
```
Comment (patrickjahns) [inline molecule/scenario_with_local_binary_directory/converge.yml:7]: As with the other scenarios, this extra converge steps shouldn't be required. Did you encounter any issues that were resolved by adding this definition?

**Comment by patrickjahns** [comment]
@OuiSouss 
Can you add your test-suite to the github actions run as well at https://github.com/patrickjahns/ansible-role-promtail/blob/master/.github/workflows/test.yml#L38-L39

So we can see the results in the pipeline run and iterate over it if there are issues

File: molecule/scenario_with_local_binary_directory/molecule.yml (hunk 1 above)
Comment (OuiSouss) [reply molecule/scenario_with_local_binary_directory/molecule.yml:15]: When using this image : 

```
  - name: instance
    image: "geerlingguy/docker-${MOLECULE_DISTRO:-ubuntu2204}-ansible:latest"
    command: ${MOLECULE_DOCKER_COMMAND:-""}
    volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:rw
    privileged: true
    cgroupns_mode: host
    pre_build_image: true
```
I encounter an issue with the test which is checking if the promtail service is running.

Maybe, it is because by systemd integration on docker and because in run molecule on a virtualbox VM.
It was my only way to found a solution to this issues but I can put the same image for the CI integration

Comment (patrickjahns) [reply molecule/scenario_with_local_binary_directory/molecule.yml:15]: 🤔  - strange - do you still have the issue at hand ? 
A while ago I had to solve some issues with systemd and the containers as well ( https://github.com/patrickjahns/ansible-role-promtail/pull/133/files#diff-90c2024f9375b6ff2916ee376a3be4f93728b19faae7a13da75e89719472390cR180 ). Maybe it's the same issue that you encountered

Comment (OuiSouss) [reply molecule/scenario_with_local_binary_directory/molecule.yml:15]: Ok, I have try to use molecule without on my host directly and not on a virtualbox VM. I confirm that it is a issue I encounter only on docker running on virtualbox.

I will push changes to use geerlingguy image

File: molecule/scenario_with_local_binary_directory/converge.yml (hunk 2 above)
Comment (OuiSouss) [reply molecule/scenario_with_local_binary_directory/converge.yml:7]: Indeed, I have just removed this file and note that it is not necessary. I thought he was. as much for me, I withdraw it.

File: tasks/install.yml [hunk 3]
```diff
@@ -63,7 +63,20 @@
     dest: "{{ promtail_tmp_dir }}/{{ promtail_version }}_promtail-linux-{{ go_arch }}.zip"
     force: True
     checksum: "{{ promtail_custom_checksum if promtail_custom_checksum else 'sha256:' + __promtail_checksum }}"
-  when: not promtail_binary.stat.exists
+  when:
+    - not promtail_binary.stat.exists
+    - promtail_binary_local_dir | length == 0
+
+- name: propagate locally distributed promtail archive
```
Comment (patrickjahns) [inline tasks/install.yml:70]: nitpick from the linter

> ERROR: [ANSIBLE0007] Standard 'Name of tasks and handlers must be formatted' not met:
> ... tasks/install.yml:70: name 'propagate locally distributed promtail archive' should start with uppercase

```suggestion
- name: Propagate locally distributed promtail archive
```

File: tasks/preflight.yml [hunk 4]
```diff
@@ -25,19 +25,26 @@
...
-- name: "Get checksum list"
-  set_fact:
-    __promtail_checksums: "{{ lookup('url', 'https://github.com/grafana/loki/releases/download/v' + promtail_version + '/SHA256SUMS', wantlist=True) | list }}"
-  run_once: True
+- name: Test if zip local archive exists
+  stat:
+    path: "{{ promtail_binary_local_dir }}/{{ promtail_version }}_promtail-linux-{{ go_arch }}.zip"
+  register: promtail_binary_local_archive
   delegate_to: localhost
+  run_once: true
```
Comment (patrickjahns) [inline tasks/preflight.yml:34]: nitpick from the linter

> ERROR: [ANSIBLE0014] Standard 'Literal bools should start with a capital letter' not met:
> ... tasks/preflight.yml:34: literal bools should be written as `True/False` or `yes/no`

```suggestion
  run_once: True
```

File: molecule/scenario_with_local_binary_directory/tests/test_default.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+"""Role testing files using testinfra."""
+
+
+def test_hosts_file(host):
+    """Validate /etc/hosts file."""
+    f = host.file("/etc/hosts")
+
+    assert f.exists
+    assert f.user == "root"
+    assert f.group == "root"
```
Comment (patrickjahns) [inline molecule/scenario_with_local_binary_directory/tests/test_default.py:10]: I believe this is a leftover file - can you please remove it from the PR

Comment (OuiSouss) [reply molecule/scenario_with_local_binary_directory/tests/test_default.py:10]: Yes i will remove it

**Comment by patrickjahns** [comment]
Everything is ✅ - thank you so much for your contribution @OuiSouss 👏

**Review by patrickjahns — APPROVED** [review]

**Comment by OuiSouss** [comment]
Cool, thanks for the review.

It was very challenging make approve this PR

**Comment by patrickjahns** [comment]
> It was very challenging make approve this PR

Do you have any feedback to share, how I can improve the process here? Is there anything that I should personally change to accommodate contributors better?

**Comment by OuiSouss** [comment]
> Do you have any feedback to share, how I can improve the process here? Is there anything that I should personally change to accommodate contributors better?

I think the process is good. You are very attentive to all notification on PR. Your comments are relevant.
One little things you can add to help contributors could be a CONTRIBUTORS.md to inform about process, test ... and an extra point, you could create a template file for PR and Issues.
